src/hotspot/share/opto/parse1.cpp

```diff
@@ -1883,12 +1883,10 @@ void Parse::merge_common(Parse::Block* target, int pnum) {
               // case #1: n is live in its block. we need to import it to AS because phi is not top().
               // references encounter before are constants including nullptr or arguments.
 
-              ObjectState* pred_os = pred_as.get_object_state(id);
-              if (pred_os->is_virtual()) {
-                as.update(id, pred_os->clone());
-              } else {
-                as.update(id, new EscapedState(phi));
-              }
+              // current block has not seen id before. it's likely m is null, n is not!
+              // we mark phi=(null, n) as 'Escaped'. or we have 'or we have Bad graph detected'.
+              //
+              as.update(id, new EscapedState(phi));
```

Comment:
This seems REALLY conservative. There is aways going to be a point where current block has not seen id before. Wouldn't this make all merged objects `phi(a,b)` escaped, whether or not it has null inputs?

Reply:
We are basically doing thing like m' = phi(m, n);  

We haven't seen `m` before in allocation state. It's possible that t is ConP, a null or a global variable(not allocate in current Compilation Unit). yes, it's conservative. we mark it 'escaped' in case we materialize m' down the road. The general rule is that m' doesn't have 'unique type' anymore. It would be error-prone to 'materialize' it, wouldn't it?  we better stop here. 

We really "miss" the chance if there's no escaping point for n down the road. however, bear in mind that we don't do scalar replacement in PEA. C2 optimizer does. it would do scalar replacement for n if it's indeed non-escaped.

Reply:
My concern with this change is that it looks like its materializing all phis, no matter what. Is there a case where a phi can still be virtual?